horovod/mxnet/compression.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2020 Amazon Inc. All Rights Reserved.
+# Copyright 2018 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Gradient compression algorithms."""
+
+import mxnet
+
+
+class Compressor(object):
+    """Interface for compressing and decompressing a given tensor."""
+    @staticmethod
+    def compress(tensor):
+        """Compresses a tensor and returns it with the context needed to decompress it."""
+        pass
+
+    @staticmethod
+    def decompress(tensor, ctx):
+        """Decompress the tensor with the given context."""
+        pass
+
+
+class NoneCompressor(Compressor):
+    """Default no-op compression."""
+    @staticmethod
+    def compress(tensor):
+        """Returns the tensor unmodified."""
+        return tensor, None
+
+    @staticmethod
+    def decompress(tensor, ctx):
+        """Returns the tensor unmodified."""
+        return tensor
+
+
+class FP16Compressor(Compressor):
+    """Compress all floating point gradients to 16-bit."""
+    @staticmethod
+    def compress(tensor):
+        """Downcasts the tensor to 16-bit."""
+        tensor_compressed = tensor
+        if 'float' in str(tensor.dtype):
+            # Only allow compression from other floating point types
+            tensor_compressed = tensor.astype('float16', copy=False)
+        return tensor_compressed, tensor.dtype
+
+    @staticmethod
+    def decompress(tensor, ctx):
+        """Upcasts the tensor to the initialization dtype."""
+        tensor_decompressed = tensor
+        dtype = ctx
+        if 'float' in str(dtype):
+            tensor_decompressed = tensor.astype(dtype, copy=False)
+        return tensor_decompressed
+
+
+class Compression(object):
+    """Optional gradient compression algorithm used during push_pull."""
```

Comment:
Change `push_pull` to `allreduce` here to be consistent with Horovod terminology.